apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs

```diff
@@ -0,0 +1,337 @@
+                    .SetupRollbackAsync();
+
+                SpannerConnection connection = BuildSpannerConnection(spannerClientMock);
+
+                var scheduler = (FakeScheduler)connection.Builder.SessionPoolManager.SpannerSettings.Scheduler;
+                var time0 = scheduler.Clock.GetCurrentDateTimeUtc();
+                var callee = new Callee(scheduler);
+
+                await scheduler.RunAsync(async () =>
+                {
+                    var result = await connection.RunWithRetriableTransactionAsync(transaction => callee.DatabaseWorkAsync(connection, transaction));
+                    Assert.Equal(3, result);
+                });
+
+                callee.AssertCallTimesInrange(
+                    new Tuple<DateTime, DateTime>(time0, time0),
+                    new Tuple<DateTime, DateTime>(time0.AddMilliseconds(SecondCallLowEx), time0.AddMilliseconds(SecondCallHighEx)),
+                    new Tuple<DateTime, DateTime>(time0.AddMilliseconds(ThirdCallLowEx), time0.AddMilliseconds(ThirdCallHighEx)));
+                Assert.InRange(scheduler.Clock.GetCurrentDateTimeUtc(), time0.AddMilliseconds(ThirdCallLowEx), time0.AddMilliseconds(ThirdCallHighEx));
+            }
+
+            [Fact]
+            public async Task CommitAbortsAlways_RespectsOverallDeadline()
+            {
+                var spannerClientMock = SpannerClientHelpers
+                    .CreateMockClient(Logger.DefaultLogger, MockBehavior.Strict)
+                    .SetupBatchCreateSessionsAsync()
+                    .SetupBeginTransactionAsync()
+                    .SetupExecuteBatchDmlAsync()
+                    .SetupCommitAsync_FailsAlways(statusCode: StatusCode.Aborted)
+                    .SetupRollbackAsync();
+
+                SpannerConnection connection = BuildSpannerConnection(spannerClientMock);
+
+                var scheduler = (FakeScheduler)connection.Builder.SessionPoolManager.SpannerSettings.Scheduler;
+                var time0 = scheduler.Clock.GetCurrentDateTimeUtc();
+                var callee = new Callee(scheduler);
+
+                await scheduler.RunAsync(async () =>
+                {
+                    var exception = await Assert.ThrowsAsync<SpannerException>(() => connection.RunWithRetriableTransactionAsync(transaction => callee.DatabaseWorkAsync(connection, transaction)));
+                    Assert.True(exception.IsRetryable && !exception.SessionExpired);
+                    Assert.Contains("Bang!", exception.InnerException.Message);
+
+                });
+
+                // The overall deadline is of 1 hour. The maximum backoff delay is of 32s.
+                // Because of jitter retries can stop at any time after 60mins - 64s, let's give it 2 minutes of range.
+                Assert.InRange(scheduler.Clock.GetCurrentDateTimeUtc(), time0.AddMinutes(58), time0.AddMinutes(60));
```

Comment:
Can we check that there's been at least 110 (or whatever the theoretical minimum is) calls?